### ISSUE TITLE
infinite loop optimized away unless function is exported

### ISSUE THREAD

**Issue by numsim1415** [ziglang/zig#1658]
```rust
export fn square(num: u8) u8 {
             
             //while(true){} // would generate loop
             return eternity(123); 
}

pub fn eternity(bla:u8) u8 {
  while(true)// is optimized away completely
  {}
}
```
See asm output on [goldbold](https://godbolt.org/z/IKsxTQ).
As you said during live coding, it seems like the eternal loop is considered to have no side effects and is thus removed.

Expected behaviour would be to refuse compilation
- as there is no return of an u8 
- as in case of `eternity` being declared as returning `void`, an implicit `return;` at the end of the function cannot be reached

**Comment by faerryn** [comment]
Are we gonna solve the halting problem?

**Comment by IridescentRose** [comment]
Isn't this a long time LLVM bug present in other languages like C and C++ where an empty infinite loop results in no consequences and thus creates invalid programs? (for example there are many articles like https://stackoverflow.com/questions/59925618/how-do-i-make-an-infinite-empty-loop-that-wont-be-optimized-away). I guess the solution Zig would need for this specific case is adding in a `nop` op-code inside of the loop so LLVM says it "has an effect" and doesn't generate an invalid program/function.

**Comment by IridescentRose** [comment]
```zig
 export fn square(num: u8) u8 {          
    //while(true){}
    return eternity(123); 
}

pub fn eternity(bla:u8) u8 {
    while(true){
        asm volatile("nop");
    }
}
```
This code generates as intended as we can see on GodBolt.

**Comment by InfRandomness** [comment]
> ```zig
>  export fn square(num: u8) u8 {          
>     //while(true){}
>     return eternity(123); 
> }
> 
> pub fn eternity(bla:u8) u8 {
>     while(true){
>         asm volatile("nop");
>     }
> }
> ```
> 
> This code generates as intended as we can see on GodBolt.

You probably have forgot to apply the flag -Drelease-safe, this happens to me as well on godbolt using Zig trunk

**Comment by Sobeston** [comment]
Related: https://github.com/rust-lang/rust/pull/82884 - apparently this was fixed in LLVM 12.

```rs
export fn square(num: u8) u8 {
    _ = num;
    return eternity(123);
}

pub fn eternity(bla: u8) u8 {
    _ = bla;
    while (true) {}
}

export fn square2(num: u8) u8 {
    _ = num;
    while (true) {}
}

export fn square3() u8 {
    while (true) {}
}
```
This code, paired with `zig build-obj` on my master build of Zig (approx 1 week old) is producing while loops correctly in all three cases, in all four -O flags.

![image](https://user-images.githubusercontent.com/15335529/134345912-d1810055-ae19-4bb8-be00-deb47916adbb.png)
(pictured above: ReleaseSmall)

```rs
pub fn main() void {
    while (true) {}
}
```
I've also tested this executable with all four -O flags, and this works on all of them.
